### PR TITLE
Sync start of auction with submission logic AND blockchain

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
Last part for https://github.com/cowprotocol/services/issues/486

# Changes
Either await new block **OR** signal from submission logic to start right away.
Will not go into details here since I wrote extensive comments to explain the somewhat subtle issue.

One open question: Do you think we should always skip waiting for the next block for the very first auction? My reasoning is that a sub-optimal auction synchronization might be better UX than having up to 12s of no action after restarts.

## How to test
Will add an e2e test using the new runmode in a separate PR
Having this untested until then seems fine since that behavior is feature gated and the changes were pretty minor so far.

### CODE REVIEW

File: crates/autopilot/src/run_loop.rs [hunk 1]
```diff
@@ -632,14 +668,14 @@ impl RunLoop {
...
         let deadline = current.saturating_add(max_blocks_wait);
         tracing::debug!(%current, %deadline, %auction_id, "waiting for tag");
         loop {
-            if self.eth.current_block().borrow().number > deadline {
-                break;
-            }
-
+            // It's impossible for the tx to make it into the current block
+            // so we start out by waiting for the next block to appear.
+            let block = ethrpc::block_stream::next_block(block_stream).await;
```
Comment (MartinquaXD) [inline crates/autopilot/src/run_loop.rs:678]: Waiting for the next block appears to get the e2e tests stuck. Not sure why, though. Investigating...

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:678]: Ahh, this is slightly annoying. Looks like the test correctly awaited the block the tx was part of but
```rust
            match self
                .persistence
                .find_tx_hash_by_auction_id(auction_id)
                .await
```
could not find the tx in the DB because it hasn't been processed, yet. :/
And since the testnet only mints blocks on demand we won't try to find the tx on the next block.
That indicates that our settlement indexing also needs to be incorporated into the runloop to ensure that all relevant events have been processed at that point.

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:678]: @sunce86 do you foresee any problems if we adjust the `on_settlement_event_updater` to run in 2 modes:
1. current mode is a background task which picks oldest unprocessed settlement event
2. "manual mode" which allows passing in which tx to process that can be called by the runloop

Are there any assumptions in the updater that events have to be processed in order?

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:678]: Looks like we need to know the `block_number` and `log_index` the settlement with an auction_id which we only have after we already indexed the "raw" settlement event. Very annoying that we don't store the `tx_hash` with every event. That would make this simpler. Let's see what can be done about this.

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:678]: @MartinquaXD I don't forsee any issues with double mode. In manual mode, given the tx hash we can easily get event `block_number` and `log_index` and do the rest the same way it's working right now.

> Very annoying that we don't store the tx_hash with every event. That would make this simpler.

Not sure I understood this one, raw SettlementEvent has (block_number, log_index, hash) or you would like to have a hash somewhere else?

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs]: > raw SettlementEvent has (block_number, log_index, hash)

True, misread the code.
However, after digging more through the code it looks like there is another issue with the double mode. The event updated needs to have the raw settlement event already indexed before it can work (at least at the moment). Because indexing the raw events happens in another separate background task it could happen that we call the updater manually but it still fails because the raw event was not yet indexed. :/

I sketched out an idea in [06ddc1a](https://github.com/cowprotocol/services/pull/2900/commits/06ddc1a5ecd3f187e75b083a978ce68f765babe3) which should already be as fast as possible but can detect timeouts incorrectly when there are multiple settlements in 1 block and our's is not the first.

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs]: I don't think we should rely on `OnSettlementEventUpdater` to be on a critical path for runloop though. It's error prone. 

If we definitely need <auction, hash> relation on a critical runloop path so we could optimize runloop submission (to end it before deadline) then I would prefer to extract building <auction, hash> into a separate task (but still do not make it part of event indexing as it contains slow web3 call).

Bottom line, we would have a COMBO condition:
1. new block received
2. events indexed, trigger 3 
3. <auction, hash> relation populated

Now use this COMBO condition in runloop instead of only (1)

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs]: What you described with your COMBO condition is effectively what my current workaround implements.

The raw event indexing listens to the block stream for updates and can notify the event updater that there is work to be done (this was there before the PR). With my workaround the event updater can also signal to the runloop that it processed something so now the runloop always waits for either the next block or the next processed settlement and reacts accordingly.
Since the whole pipeline relies on message passing there should be minimal latency between indexing a new block and the runloop learning about the fully processed settlement.

It's not really pretty IMO but I think this is effectively what has to happen without requiring a probably big refactoring upfront.

File: crates/autopilot/src/run_loop.rs [hunk 2]
```diff
@@ -98,6 +91,47 @@ impl RunLoop {
...
+            // transaction submission logic that tells us to start right away and **not**
+            // wait for the next block.
+            let await_new_block = ethrpc::block_stream::next_block(self.eth.current_block());
+            pin_mut!(await_new_block);
+            let await_notify = self.run_loop_start_notifier.notified();
+            pin_mut!(await_notify);
+            futures::future::select(await_notify, await_new_block).await;
+
+            let block_number = self.eth.current_block().borrow().number;
+            if let Err(err) = self.solvable_orders_cache.update(block_number).await {
```
Comment (sunce86) [inline crates/autopilot/src/run_loop.rs:125]: I think we also need to make sure our COMBO condition is satisfied before building auction here. But this could also be an optimization over current implementation, not necessarily blocking.

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:125]: Why should the COMBO condition from the transaction detection discussion be relevant here?
But regardless the tx detection code await the COMBO condition and if we reached the deadline it informs this code to NOT wait for the next block and instead start right away.
This is how it's supposed to be from my understanding or am I missing something?

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:125]: > Why should the COMBO condition from the transaction detection discussion be relevant here?

So that auction is built on top of executed trades from the newest block. DB needs Trade events indexed so it can update the state of orders and return only OPEN orders to SolvableOrdersCache. Then we can also drop inflight orders. But as I said, I am not sure this would be a requirement for this PR.

Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:125]: AFAIU unless we run into the issue discussed [here](https://github.com/cowprotocol/services/pull/2900#discussion_r1719591771) we should be okay. Will think more about avoiding that issue.

File: crates/autopilot/src/run_loop.rs [hunk 3]
```diff
@@ -695,7 +741,13 @@ impl RunLoop {
                 }
                 Ok(None) => {}
             }
-            tokio::time::sleep(Duration::from_secs(3)).await;
+            if block_number == deadline {
```
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:744]: always inequality here please.

File: crates/driver/src/infra/api/routes/settle/mod.rs [hunk 4]
```diff
@@ -16,18 +19,28 @@ async fn route(
     state: axum::extract::State<State>,
     solution: axum::Json<dto::Solution>,
 ) -> Result<(), (hyper::StatusCode, axum::Json<Error>)> {
-    let competition = state.competition();
-    let auction_id = competition.auction_id().map(|id| id.0);
-    let handle_request = async {
+    let state = state.clone();
```
Comment (fleupold) [inline crates/driver/src/infra/api/routes/settle/mod.rs:22]: Good catch! This should probably be a standalone fix...

File: crates/autopilot/src/on_settlement_event_updater.rs [hunk 5]
```diff
@@ -55,6 +56,10 @@ struct Inner {
     persistence: infra::Persistence,
     db: Postgres,
     notify: Notify,
+    /// Can be monitored to get notified about processed events.
+    updates_receiver: watch::Receiver<()>,
+    /// Can be used to notify about new processed events.
+    updates_sender: watch::Sender<()>,
```
Comment (fleupold) [inline crates/autopilot/src/on_settlement_event_updater.rs:62]: Having these different senders and receivers being subscribed from and triggered by very different parts of the code base makes things very hard to reason about imo (e.g. it's not clear to me what would happen if the settlement was included before the deadline, would we wait for another block in the runloop before starting a new auction?).

Can't we make everything synchronous in the runloop (and have the only condition that  on the block number must have changed compared to the last runloop)? Is it really necessary to have all these updates happen in a separate tasks that need trigger different synchronisation barriers? 

The runloop could imo simply be:
1. If there has been a previous auction with block x, wait for block x it to change (will most often be immediately the case)
2. [update_events](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/shared/src/event_handling.rs#L254), this ensures all settlements are processed for all blocks until the current one.
3. [update](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/autopilot/src/solvable_orders.rs#L169) the SolvableOrdersCache (this creates an auction with latest state applied)
4. Let solvers solve + winner settle
5. Repeat...

In terms of latency this should be the same as what is proposed here (we are doing the same amount of async work) but the code path will be much simpler to reason about.

**Review by fleupold — COMMENTED** [review]
I think we should make almost everything that is needed for data consistency synchronously updated in our new run loop mode.

File: crates/autopilot/src/on_settlement_event_updater.rs (hunk 5 above)
Comment (MartinquaXD) [reply crates/autopilot/src/on_settlement_event_updater.rs:62]: This has a problem when a solver wins an auction and then returns from the `/settle` call that it stopped trying the solution on-chain. In that case the start of the next auction might happen in the middle between 2 blocks. But that can also be solved with local variables (just thought about that 🤦 )
Overall I agree that it would be ideal to not jump through the hoops with the `OnEventUpdater` signal but I'm worried that this would be a significant refactor which would delay getting valuable performance measurements that could be worked on in parallel.

Indexing events becomes a lot less straight forward because we still need to be able to index historic events. One idea how to deal with I sketched out [here](https://github.com/cowprotocol/services/pull/2900#discussion_r1719448355) but this doesn't really work since it can leave the DB in an inconsistent state when there are historic events to be indexed, but the runloop already indexed a few more current events and the service getting restarted.
In that case we'd have a gap of unindexed events but we would not reindex this gap on the next restart because we already have more recent events indexed and we always start at the latest block we have any event for.

After thinking about this a bit more to me it seems indexing would need to be redesigned with these requirements:
1. on restart index all missing historic events (delays first run loop but shouldn't be too bad)
2. while waiting for the tx to land index all events we missed in the meantime
3. find tx on-chain and start next loop

Comment (fleupold) [reply crates/autopilot/src/on_settlement_event_updater.rs:62]: > In that case the start of the next auction might happen in the middle between 2 blocks.

Ok, so instead of the only condition being the block changed, we can also say something like "less than x seconds passed since the current block's timestamp". This way, if we would start an auction too late we rather skip and wait to start the next one on time.

 > Indexing events becomes a lot less straight forward because we still need to be able to index historic events.

Currently indexing events is [triggered](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/shared/src/event_handling.rs#L506) from a maintenance task (similar to how solvable order updates is triggered from a background task). Why can we not in case of mode `SyncToBlockchain` call `update_events` from within the runloop? This will index all historic events and avoid ending up in inconsistent state. Most of the time `update_events` only processes a single block. In the case of a backlog there won't be any auctions but this might be ok?

**Comment by MartinquaXD** [comment]
Closed in favor of https://github.com/cowprotocol/services/pull/2906
